Add writable flag to meta stores

kiksekage opened the pull request:
Fixes #255

The `MetaStore` class is set to be writable by default:
https://github.com/DHI-GRAS/terracotta/blob/a3aedcbfcd95aed279dd9754434b6409462b4bc1/terracotta/drivers/base_classes.py#L41

Added a very thin decorator to throw the new `DatabaseNotWritable` exception:
https://github.com/DHI-GRAS/terracotta/blob/a3aedcbfcd95aed279dd9754434b6409462b4bc1/terracotta/drivers/terracotta_driver.py#L23-L34

And then we access the flag directly when getting metadata from the meta store:
https://github.com/DHI-GRAS/terracotta/blob/22353ca9c3f99dbe4cdb70b7497d93162827713a/terracotta/drivers/terracotta_driver.py#L187-L194

Me and @nickeopti were discussing whether or not it was cleanest to handle the database "writability" in `TerracottaDriver` or all meta stores individually.

dionhaefner commented:
I like introducing a new exception. But I think it would be cleaner if the raster drivers themselves would check for writability and raise the exception if necessary, and we can just let that bubble up to the client. Then we can also have this:

```python
class RemoteSQLiteRasterStore:
    @requires_writable
    def insert(self, *args, **kwargs):
        pass  # always unwritable
```

In this case `writable` could be a private attribute `self._writable`, and all interactions could happen via this exception. For example, to raise a better error in conjunction with lazy loading we could catch the exception and re-raise it with a different message:

```python
try:
    self.insert(keys, path, metadata=metadata) 
except DatabaseNotWritable as exc:
    raise DatabaseNotWritable("Lazy loading requires a writable database") from exc
```

kiksekage commented:
I moved the handling into the meta stores instead, with the explicit catching as well. If it looks good to you @dionhaefner I'll write some tests to up the coverage 👍

dionhaefner commented on terracotta/drivers/sqlite_remote_meta_store.py:
```diff
@@ -132,14 +135,17 @@ def _connection_callback(self) -> None:
         self._update_db(self._remote_path, self._local_path)
         super()._connection_callback()
 
+    @requires_writable
```
Please add a comment that this always raises

dionhaefner commented on terracotta/drivers/terracotta_driver.py:
```diff
@@ -169,11 +169,17 @@ def get_metadata(self, keys: ExtendedKeysType) -> Dict[str, Any]:
-            metadata = self.meta_store.get_metadata(keys)
-            assert metadata is not None
+            try:
+                self.insert(keys, path, metadata=metadata)
```
Only this line should be in `try-except`

dionhaefner commented on terracotta/exceptions.py:
```diff
@@ -24,5 +24,9 @@ class InvalidDatabaseError(Exception):
     pass
 
 
+class DatabaseNotWritable(Exception):
```
Should be `DatabaseNotWritableError`

dionhaefner reviewed (commented):
Yes that looks pretty good. Thanks!

I think we will also need an error handler for the new exception but the tests should reveal that.

BTW, can we check whether the other drivers are writable via SQLalchemy? It should just boil down to checking whether the current user has write permissions on the used database.

kiksekage commented:
The failing macos tests go through now and the PR should be ready to merge if you have the time @dionhaefner @mrpgraae 👍

dionhaefner commented:
> I think we will also need an error handler for the new exception but the tests should reveal that.

> BTW, can we check whether the other drivers are writable via SQLalchemy? It should just boil down to checking whether the current user has write permissions on the used database.

Could you follow up on these points? What's the status?

kiksekage commented:
> Could you follow up on these points? What's the status?

Good point, I will have a look at that.

kiksekage commented:
> BTW, can we check whether the other drivers are writable via SQLalchemy? It should just boil down to checking whether the current user has write permissions on the used database.

I can't seem to find any ORM/core way of doing this in sqlalchemy, though I might be looking in the wrong directions. [This](https://github.com/sqlalchemy/sqlalchemy/issues/4239) and [this](https://github.com/sqlalchemy/sqlalchemy/issues/3326) (very old) issue seem to indicate that interfacing with user privileges is done the raw SQL query way, and then we are back to having to discern between which backend we are using and what syntax to use.
I guess we could do a write to any/all of the four tables and a rollback at `meta_store` instantiation time, but that also doesn't sit right with me.

> I think we will also need an error handler for the new exception but the tests should reveal that.

I confused myself, does this refer to adding `exceptions.DatabaseNotWritableError` to the API error handlers here: https://github.com/DHI-GRAS/terracotta/blob/b5ee523be89dba5a622633751f4181ad7f827246/terracotta/server/flask_api.py#L41

dionhaefner commented:
Too bad with the permissions. Maybe doing it in SQL is not so bad? But then again it might not be portable between different DB vendors? Your call.

> I confused myself, does this refer to adding exceptions.DatabaseNotWritableError to the API error handlers here: 

Yes, exactly. Right now this will throw a 500 internal server error, but maybe we would want something more descriptive.

nickeopti reviewed (commented):
> > I think we will also need an error handler for the new exception but the tests should reveal that.

> I confused myself, does this refer to adding exceptions.DatabaseNotWritableError to the API error handlers here: 

Yeah, that will need a flask error handler, like the ones you referenced, @kiksekage. However, it seems that we don't have any tests for lazy-loading metadata on API/server level. I guess we should add that. I must admit I haven't quite grasped how that setup works, so I haven't added one.

dionhaefner commented on terracotta/drivers/terracotta_driver.py:
```diff
@@ -169,7 +169,13 @@ def get_metadata(self, keys: ExtendedKeysType) -> Dict[str, Any]:
             metadata = self.compute_metadata(path, max_shape=self.LAZY_LOADING_MAX_SHAPE)
-            self.insert(keys, path, metadata=metadata)
+
+            try:
+                self.insert(keys, path, metadata=metadata)
+            except exceptions.DatabaseNotWritableError as exc:
+                raise exceptions.DatabaseNotWritableError(
+                    "Lazy loading requires a writable database"
+                ) from exc
```
Maybe, if you have a convincing use case. I don't really want to encourage lazy loading too much, especially not for production contexts. So the functionality is limited by design.

kiksekage replied:
> And also, @kiksekage, the STAC driver you were working on, required lazy loading without writable, right?

Yeah, the choice was either a) maintain a local DB for metadata calculation results or b) only do on-the-fly metadata calculations, and b) seemed nicer by simply not having to maintain a local DB. This might still change though! 
I like the idea of a warning, and currently I guess it would only be raised when lazy loading on a sqlite-remote metastore. Since that didn't work actually work in the codebase before, it might be a small enough niche that it shouldn't affect too many users?

dionhaefner replied:
I suggest we keep it as it is for now and discuss this in a separate issue.

kiksekage commented on terracotta/drivers/sqlite_remote_meta_store.py:
```diff
@@ -132,14 +135,18 @@ def _connection_callback(self) -> None:
+    # Always raises DatabaseNotWritableError
+    @requires_writable
     def create(self, *args: Any, **kwargs: Any) -> None:
-        raise NotImplementedError('Remote SQLite databases are read-only')
+        pass
 
+    @requires_writable
     def insert(self, *args: Any, **kwargs: Any) -> None:
-        raise NotImplementedError('Remote SQLite databases are read-only')
+        pass
 
+    @requires_writable
     def delete(self, *args: Any, **kwargs: Any) -> None:
-        raise NotImplementedError('Remote SQLite databases are read-only')
+        pass
```
Yeah I agree, simply `pass`ing doesn't help the understanding anymore than what is already conveyed.

kiksekage commented:
> Too bad with the permissions. Maybe doing it in SQL is not so bad? But then again it might not be portable between different DB vendors? Your call.

I think, with all the work that went into sqlachemy-fying the DB handling it hit me the wrong way to start doing some vendor-specific things again. I'm in favor of not checking it and letting it be up to the user.

I have added an error handler, and as far as I could tell, the only way for a API user to hit this error would be to request metadata for a dataset which is present but doesn't have any pre-calculated metadata. The test uses the already existing fixtures for generating `sqlite-remote` databases for testing so I moved that functionality to `conftest.py`.

dionhaefner commented:
> I think, with all the work that went into sqlachemy-fying the DB handling it hit me the wrong way to start doing some vendor-specific things again. I'm in favor of not checking it and letting it be up to the user.

Sounds like sunk cost fallacy to me :) But it's fine for now, the error from SQLalchemy should be descriptive enough to alert users as to what's going on.

kiksekage commented on tests/server/test_flask_api.py:
```diff
@@ -39,6 +53,57 @@ def test_get_metadata(client, use_testdb):
+
+    rv = client.get('/metadata/some/value')
+    assert rv.status_code == 403
+
+
+@moto.mock_s3
+def test_debug_errors(debug_client, s3_db_factory, mock_aws_env, raster_file, raster_file_xyz):
+    import terracotta
+    from terracotta import exceptions
+    import marshmallow
+
+    keys = ('some', 'keys')
+    dbpath = s3_db_factory(
+        keys, datasets={("some", "value"): str(raster_file)}, skip_metadata=True
+    )
+    terracotta.update_settings(DRIVER_PATH=str(dbpath), DRIVER_PROVIDER="sqlite-remote")
+
+    with pytest.raises(exceptions.DatabaseNotWritableError):
+        debug_client.get('/metadata/some/value')
+
+    with pytest.raises(exceptions.DatasetNotFoundError):
+        debug_client.get('/metadata/NONEXISTING/KEYS')
+
+    with pytest.raises(exceptions.InvalidKeyError):
+        debug_client.get('/metadata/ONLYONEKEY')
+
+    x, y, z = raster_file_xyz
+
+    with pytest.raises(exceptions.InvalidArgumentsError):
+        debug_client.get(
+            f'/compute/some/value/{z}/{x}/{y}.png'
+            '?expression=v1*v2&v1=val22&v2=val23'
+            '&stretch_range=[10000,0]'
+        )
+
+    with pytest.raises(marshmallow.ValidationError):
+        debug_client.get(
+            f'/compute/some/value/{z}/{x}/{y}.png'
+            '?stretch_range=[10000,0]'
+        )
```
Makes sense! Will do.

dionhaefner commented on tests/conftest.py:
```diff
@@ -350,43 +345,19 @@ def use_testdb(testdb, monkeypatch):
+    with driver.connect():
+        driver.insert(('first', 'second', 'third'), str(raster_file), skip_metadata=True)
 
+    driver.meta_store._WRITABLE = False
+    yield
+    driver.meta_store._WRITABLE = True
```
You already give monkeypatch as a fixture, so why not use it? :smile: 

```suggestion
    with monkeypatch.context() as m:
        m.setattr(driver.meta_store, "_WRITABLE", False)
        yield
```

(no need to set it back)

kiksekage replied:
Huh, TIL! :clap:

dionhaefner commented on tests/drivers/test_raster_drivers.py:
```diff
@@ -144,6 +146,22 @@ def test_lazy_loading(driver_path, provider, raster_file):
+    db.create(keys)
+    db.insert(['some', 'value'], str(raster_file), skip_metadata=True)
+
+    # Manually set the meta store to un-writable
+    db.meta_store._WRITABLE = False
+
+    with pytest.raises(exceptions.DatabaseNotWritableError):
+        db.get_metadata(['some', 'value'])
```
I think this will need a monkeypatch context, too.

kiksekage replied:
True, added it now.

dionhaefner reviewed (approved):
LGTM! Thanks @kiksekage

kiksekage commented:
> LGTM! Thanks @kiksekage

Should I do the honors of merging or will you @dionhaefner? 😄
